solidity/contracts/l2/L2WormholeGateway.sol

```diff
@@ -0,0 +1,371 @@
+        // after ITokenBridge call. This way, we are sure this contract received
+        // Wormhole tBTC token in the given amount. This is transparent to
+        // all potential upgrades of ITokenBridge implementation and no other
+        // validations are needed.
+        uint256 balanceBefore = bridgeToken.balanceOf(address(this));
+        bytes memory encoded = bridge.completeTransferWithPayload(encodedVm);
+        uint256 balanceAfter = bridgeToken.balanceOf(address(this));
+
+        uint256 amount = balanceAfter - balanceBefore;
+        // Protect against the custody of irrelevant tokens.
+        require(amount > 0, "No tBTC transferred");
+
+        address receiver = fromWormholeAddress(
+            bytes32(bridge.parseTransferWithPayload(encoded).payload)
+        );
+        require(receiver != address(0), "0x0 receiver not allowed");
+
+        // We send wormhole tBTC OR mint canonical tBTC. We do not want to send
+        // dust. Sending wormhole tBTC is an exceptional situation and we want
+        // to keep it simple.
+        if (mintedAmount + amount > mintingLimit) {
+            bridgeToken.safeTransfer(receiver, amount);
+        } else {
+            // The function is non-reentrant given bridge.completeTransferWithPayload
+            // call that does not allow to use the same VAA again.
+            // slither-disable-next-line reentrancy-benign
+            mintedAmount += amount;
+            tbtc.mint(receiver, amount);
+        }
+
+        // The function is non-reentrant given bridge.completeTransferWithPayload
+        // call that does not allow to use the same VAA again.
+        // slither-disable-next-line reentrancy-events
+        emit WormholeTbtcReceived(receiver, amount);
+    }
+
+    /// @notice Allows to deposit Wormhole tBTC token in exchange for canonical
+    ///         tBTC. Useful in a situation when user received wormhole tBTC
+    ///         instead of canonical tBTC. One example of such situation is
+    ///         when the minting limit was exceeded but the user minted anyway.
+    /// @dev Requirements:
+    ///      - The sender must have at least `amount` of the Wormhole tBTC and
+    ///        it has to be approved for L2WormholeGateway.
+    /// @param amount The amount of Wormhole tBTC to deposit.
+    function depositBridgeToken(uint256 amount) external {
+        bridgeToken.transferFrom(msg.sender, address(this), amount);
+        tbtc.mint(msg.sender, amount);
+        emit WormholeTbtcDeposited(msg.sender, amount);
+    }
+
+    /// @notice Lets the governance to update the tBTC gateway address on the
+    ///         chain with the given Wormhole ID.
+    /// @param chainId Wormhole ID of the chain.
+    /// @param gateway Address of tBTC gateway on the given chain.
+    function updateGatewayAddress(uint16 chainId, address gateway)
+        external
+        onlyOwner
+    {
+        bytes32 wormholeAddress = toWormholeAddress(gateway);
```

Comment:
Is there a plan to deploy tBTC contracts on non-evm chains? If so, `updateGatewayAddress` should take a 32 byte `gateway` argument.

Reply:
If you guys aren't planning to write a non-evm specific implementation (and deploy canonical tBTC on a non-evm chain) then this should work. However, if you decide to write a Solana contract (for example) the gateway address would have to be passed as bytes32 (all Solana addresses are 32 bytes).

Reply:
That makes sense, thanks for the explanation. See https://github.com/keep-network/tbtc-v2/pull/570/commits/e6b29854de931e6473076ca4f52a953aba7210bd, please.